### PR TITLE
Block Misconfigured Peers

### PR DESCRIPTION
**Pull request by fosskers**

### CODE REVIEW

File: src/Chainweb/HostAddress.hs [hunk 1]
```diff
@@ -348,6 +357,57 @@ localhostIPv6 :: Hostname
...
+isPrivateHostname :: Hostname -> Bool
+isPrivateHostname (HostnameIPv4 ip) = isPrivateIp (read $ B8.unpack $ CI.original ip)
+isPrivateHostname h
+    | h == localhost = True
+    | h == localhostIPv4 = True
+    | h == localhostIPv6 = True
+    | otherwise = False
+
+isReservedHostname :: Hostname -> Bool
+isReservedHostname (HostnameIPv4 ip) = isReservedIp (read $ B8.unpack $ CI.original ip)
```
Comment (gregorycollins) [inline src/Chainweb/HostAddress.hs]: also needs `|| isPrivateIp ...`

File: src/P2P/Node.hs [hunk 2]
```diff
@@ -288,6 +300,38 @@ setInactive node = writeTVar (_p2pNodeActive node) False
...
+    me :: Maybe PeerId
+    me = _peerId $ _p2pNodePeerInfo node
+
+    isKnown :: PeerSet -> Bool
+    isKnown peers = not . IXS.null $ IXS.getEQ (_peerAddr pinf) peers
+
+    -- TODO Check more IP ranges, possibly in HostAddress module.
+    isTrivial :: Bool
+    isTrivial = case v of
+        Mainnet01 -> isReservedHostAddress (_peerAddr pinf)
```
Comment (gregorycollins) [inline src/P2P/Node.hs]: Either fully subsume private within reserved or add private here

File: src/Chainweb/CutDB.hs [hunk 3]
```diff
@@ -590,24 +591,28 @@ cutHashesToBlockHeaderMap
...
-        (headers :> missing) <- S.each (HM.toList $ _cutHashes hs)
-            & S.map (fmap snd)
-            & S.mapM (tryGetBlockHeader hdrs plds)
-            & S.partitionEithers
-            & S.fold_ (\x (cid, h) -> HM.insert cid h x) mempty id
-            & S.fold (\x (cid, h) -> HM.insert cid h x) mempty id
-        if null missing
-            then return $! Right headers
-            else return $! Left missing
+cutHashesToBlockHeaderMap logfun headerStore payloadStore hs = timeout 3000000 go >>= \case
```
Comment (larskuhtz) [inline src/Chainweb/CutDB.hs]: This is the cut db timeout

Comment (slpopejoy) [inline src/Chainweb/CutDB.hs]:
```suggestion
cutHashesToBlockHeaderMap logfun headerStore payloadStore hs = timeout 3000000 go >>= \case -- TODO define as constant or make configurable
```

File: src/Chainweb/Utils.hs [hunk 4]
```diff
@@ -1158,3 +1164,26 @@ threadDelayRng = unsafePerformIO (Prob.createSystemRandom >>= newMVar)
 approximateThreadDelay :: Int -> IO ()
 approximateThreadDelay d = withMVar threadDelayRng (approximately d)
                            >>= threadDelay
+
+-- -------------------------------------------------------------------------- --
+-- TLS Manager with connection timeout
+
```
Comment (slpopejoy) [inline src/Chainweb/Utils.hs]:
```suggestion
  -- TODO unify with other HTTP managers
```

File: src/P2P/Node.hs [hunk 5]
```diff
@@ -573,6 +619,7 @@ p2pCreateNode cv nid peer logfun db mgr session = do
     statsVar <- newTVarIO emptyP2pNodeStats
     rngVar <- newTVarIO =<< R.newStdGen
     activeVar <- newTVarIO True
+    cmgr <- Chainweb.Utils.manager 250000
```
Comment (slpopejoy) [inline src/P2P/Node.hs]:
```suggestion
    cmgr <- Chainweb.Utils.manager 250000 -- constant is 250milli request timeout
```

File: src/P2P/Node/RestAPI/Server.hs [hunk 6]
```diff
@@ -102,7 +104,11 @@ peerPutHandler
     -> NetworkId
     -> PeerInfo
     -> Handler NoContent
-peerPutHandler db nid e = liftIO $ NoContent <$ peerDbInsert db nid e
+peerPutHandler db nid e
```
Comment (slpopejoy) [inline src/P2P/Node/RestAPI/Server.hs]:
```suggestion
peerPutHandler db nid e -- TODO consider connection test here for bad peer
```

**Pull request merged**